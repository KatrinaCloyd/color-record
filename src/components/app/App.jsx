/* eslint-disable max-len */
import React, { useEffect, useState } from 'react';

const useRecord = () => {
  const [current, setCurrent] = useState(0);
  const [colorHistory, addToHistory] = useState(['#FF0000']);

  useEffect(() => {
    console.log('current index - useEffect', current);
    console.log('current history array - useEffect', colorHistory);
  }, [current]);

  const undo = () => {
    setCurrent((prev) => prev - 1);
  };

  const redo = () => {
    setCurrent((prev) => prev + 1);
  };

  // const record = (clr) => {
  //   console.log('chosen color', clr);
  //   console.log('current index', current);
  //   //this just adds it to the end, we need to add it in at the index given.
  //   addToHistory((prev) => [...prev, clr]);
  //   console.log('current history array', colorHistory);
  //   setCurrent((prev) => prev + 1);
  //   console.log('current index', current);
  // };

  const record = (clr) => {
    if (colorHistory.length === current + 1) {
      addToHistory((prev) => [...prev, clr]);
      setCurrent((prev) => prev + 1);
    }
    else {
      addToHistory((prevArr) => [
        ...prevArr.slice(0, current + 1), // ...before
        clr,
        ...prevArr.slice(current + 1)]); // ...after
      setCurrent((prev) => prev + 1);
    }
  };

  return { current, undo, redo, record, colorHistory };
};

function App() {

  const { current, undo, redo, record, colorHistory } = useRecord();

  return (
    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
      <div>
        <button onClick={undo} style={{ margin: '10px' }} disabled={current <= 0}>undo</button>
        <button onClick={redo} style={{ margin: '10px' }} disabled={current >= colorHistory.length - 1 || colorHistory.length <= 1}>redo</button>
      </div>
      <label /> Choose A Color:
      <input type="color" style={{ margin: '10px' }} value={colorHistory[current]} onChange={(e) => record(e.target.value)} />
      <div style={{ backgroundColor: colorHistory[current], width: '10rem', height: '10rem', margin: '10px' }}></div>
    </div>
  );
}

export default App;
