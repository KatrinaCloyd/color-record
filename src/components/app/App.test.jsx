import React from 'react';
import { getByRole, render, screen } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import App from './App';

describe('App component', () => {
  it('renders App', () => {
    const app = render(<App />);
    expect(app).toMatchSnapshot();
    //on load
    const undo = screen.getByRole('button', { name: 'undo' });
    expect(undo).toHaveAttribute('disabled');
    const redo = screen.getByRole('button', { name: 'redo' });
    expect(redo).toHaveAttribute('disabled');
    const colorInput = screen.getByLabelText('Choose A Color:');
    expect(colorInput).toHaveValue('#ffffff');

    // fireEvent.input(colorInput, { target: { value: '#FF0000' } });

  });
});

// select #FF0000 
// Expect the square change to #FF0000
// select #0000FF 
// Expect the square change to #0000FF
// select #00FF00
// Expect the square change to #00FF00
// click undo
// Expect the square change to #0000FF
// press undo
// Expect the square change to #FF0000
// press redo
// Expect the square change to #0000FF
// select #FFFF00
// Expect the square change to #FFFF00
// press undo
// Expect the square change to #0000FF
// press undo
// Expect the square change to #FF0000
// press redo
// Expect the square change to #0000FF
// press redo
// Expect the square change to #FFFF00
// press redo
// Expect the square change to #00FF00
