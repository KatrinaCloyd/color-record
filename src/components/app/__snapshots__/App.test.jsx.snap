// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: flex; flex-direction: column; align-items: center;"
      >
        <div>
          <button
            aria-label="undo"
            disabled=""
            style="margin: 10px;"
          >
            undo
          </button>
          <button
            aria-label="redo"
            disabled=""
            style="margin: 10px;"
          >
            redo
          </button>
        </div>
        <label>
           Choose A Color:
          <input
            aria-label="color-input"
            style="margin: 10px;"
            type="color"
            value="#ffffff"
          />
        </label>
        <div
          style="background-color: rgb(239, 239, 239); border: .25px solid #919191; border-radius: 3px; width: 11rem; height: 11rem; display: flex; flex-direction: column; align-items: center; justify-content: center;"
        >
          <div
            aria-label="clr-box"
            style="background-color: rgb(255, 255, 255); border: .25px solid #919191; width: 10rem; height: 10rem;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: flex; flex-direction: column; align-items: center;"
    >
      <div>
        <button
          aria-label="undo"
          disabled=""
          style="margin: 10px;"
        >
          undo
        </button>
        <button
          aria-label="redo"
          disabled=""
          style="margin: 10px;"
        >
          redo
        </button>
      </div>
      <label>
         Choose A Color:
        <input
          aria-label="color-input"
          style="margin: 10px;"
          type="color"
          value="#ffffff"
        />
      </label>
      <div
        style="background-color: rgb(239, 239, 239); border: .25px solid #919191; border-radius: 3px; width: 11rem; height: 11rem; display: flex; flex-direction: column; align-items: center; justify-content: center;"
      >
        <div
          aria-label="clr-box"
          style="background-color: rgb(255, 255, 255); border: .25px solid #919191; width: 10rem; height: 10rem;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
